### ISSUE TITLE
Switch scene and keep the play node

### ISSUE THREAD

**Issue by AlexHolly** [godotengine/godot#2657]
http://www.godotengine.org/topics/12346

Still not solved, switching the scene and keeping a node(player) is essential and should work out of the box in 2.0.

If you need a code example, please tell me I will try to post some next week.

**Comment by vnen** [comment]
This is not really an issue (or rather: you're not describing the _actual_ issue).

Let me be clear: the issue is that the `_ready()` function should not be called when the node reenter the tree. However, the current functionality still might be useful, as it's the only reliable way to know that not only the node entered the tree but also all of its children too.

So, what you need is equivalent of `_ready()` that is guaranteed to run only once. `_init()` doesn't do it because it's actually called before the node enter the tree. To get a chance in being answered, I suggest you ask for that (which is a general issue that other people might be having) instead of what you're currently asking (which is a specific issue to you project and by being so has many solutions that does not affect the engine).

The easiest way around your problem (and has been already suggested in the forum thread) is to use a flag to see if it's already initialized so you don't do it more than once:

```GDScript
var initialized = false
func _ready():
	if !initialized:
		initialized = true
		# Do initialization stuff
```

Another solution is to keep the scene always open and remove the children that you don't need (that level) while keeping the ones you do need (the player). This might seem complicated to change form an existing project but not hard to do from the beginning.

**Comment by volzhs** [comment]
I agree with @vnen :+1:

**Comment by AlexHolly** [comment]
I've added the initialized code to the player and all children and I still get 3 error messages.

It is still a workaround even if it would work. 
Let's say I missed one child who is throwing those 3 errors and the workaround works.

Adding that initialize code to the player and all of his children manually shows that a function is needed.

As I already mentioned something like the function below would safe a lot code and time. 

``` add_child(node, initialized=false) ```

Is there any reason why this function is a bad idea?
Will it break something?
Is it impossible to implement?

Thanks for the quick response.

**Comment by volzhs** [comment]
I think, it will not break something but more flexibility causes more complexity.

Engine should report error if programmer does something wrong.
Engine can not control properly all cases as what programmer actually wants.

**Comment by volzhs** [comment]
if godot implements ``add_child(node, initialized=false)``,
there will be somebody asks "there is ``add_child(node, initialized=false)`` and why we need ``func _init()``?"

**Comment by bojidar-bg** [comment]
BTW, maybe `initialize=false` isn't the best solution: http://ariya.ofilabs.com/2011/08/hall-of-api-shame-boolean-trap.html
I would purpose to instead have `add_child_without_initializing` or something along those lines.

**Comment by AlexHolly** [comment]
I've seen this default stuff on one or two places in Godot so I thought it's an standart. 
I don't like it either but it makes the api smaller.

**Comment by reduz** [comment]
I think it's pointless to add something like this. It's a special case so
it should be handled as such, just keep a boolean and mark it as
initialized.

Re Initialization when moving from scene to scene may be desirable
depending on the context, the node, etc. It's best if this is handled
manually and not as an engine feature.

**Comment by vnen** [comment]
@AlexHolly I'm proposing to close this issue, as it does not describe an engine feature and it's full of noise. If you have a feature to request, make another issue with a clear description of the changes in the engine. If you think there's an actual bug, again make another issue with clear description and steps to reproduce (possibly with a sample project).

**Comment by AlexHolly** [comment]
Finally I found the problem.
I made an example project https://www.dropbox.com/s/2jix96clogql0p9/SwitchScene.zip?dl=0
You can switch the scene with left and right arrow.
Look at the errors I uploaded it with the bug.
To fix it you have to go to "ActionArea" and disable Monitoring.

I suggest to improve the debug informations to show which node caused the problem,
I don't know if it is possible to show the Node name as Object has no get_name() but it is possible to show the object id. So I suggest to improve the debugging like this.

core/object.cpp -> line 1376 -> ERR_EXPLAIN("Signal '"+p_signal+"'' already connected to given method '" + p_to_method + "' in that object. ID: " + itos(target._id));

Thanks for you help.
